src/app/features/ledger/generic-flows/tx-signing/use-ledger-sign-tx.ts

```diff
@@ -0,0 +1,102 @@
+import { useMemo, useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+
+import StacksApp from '@zondax/ledger-stacks';
+import BitcoinApp from 'ledger-bitcoin';
+
+import { SupportedBlockchains } from '@shared/constants';
+import { RouteUrls } from '@shared/route-urls';
+
+import { delay } from '@app/common/utils';
+
+import { useLedgerNavigate } from '../../hooks/use-ledger-navigate';
+import { BitcoinAppVersion } from '../../utils/bitcoin-ledger-utils';
+import {
+  LedgerConnectionErrors,
+  checkLockedDeviceError,
+  useLedgerResponseState,
+} from '../../utils/generic-ledger-utils';
+import {
+  StacksAppVersion,
+  isVersionOfLedgerStacksAppWithContractPrincipalBug,
+} from '../../utils/stacks-ledger-utils';
+import { createWaitForUserToSeeWarningScreen } from './ledger-sign-tx.context';
+
+interface UseLedgerSignTxArgs<App extends BitcoinApp | StacksApp> {
+  chain: SupportedBlockchains;
+  isAppOpen({ name }: { name: string }): boolean;
+  getAppVersion(app: App): Promise<StacksAppVersion> | Promise<BitcoinAppVersion>;
+  connectApp(): Promise<App>;
+  onSuccess?(): void;
+  signTransactionWithDevice(app: App): Promise<void>;
+}
+
+export function useLedgerSignTx<App extends StacksApp | BitcoinApp>({
+  chain,
+  isAppOpen,
+  getAppVersion,
+  connectApp,
+  onSuccess,
+  signTransactionWithDevice,
+}: UseLedgerSignTxArgs<App>) {
+  const [outdatedAppVersionWarning, setAppVersionOutdatedWarning] = useState(false);
+  const [latestDeviceResponse, setLatestDeviceResponse] = useLedgerResponseState();
+  const [awaitingDeviceConnection, setAwaitingDeviceConnection] = useState(false);
+  const ledgerNavigate = useLedgerNavigate();
+  const hasUserSkippedBuggyAppWarning = useMemo(() => createWaitForUserToSeeWarningScreen(), []);
+  const navigate = useNavigate();
+  async function checkCorrectAppIsOpenWithFailState(app: App) {
+    const response = await getAppVersion(app);
+    if (!isAppOpen({ name: response.name })) {
+      setAwaitingDeviceConnection(false);
+      throw new Error(LedgerConnectionErrors.AppNotOpen);
+    }
+    if (response.chain === 'stacks') {
+      if (isVersionOfLedgerStacksAppWithContractPrincipalBug(response)) {
+        navigate(RouteUrls.LedgerOutdatedAppWarning);
+        const response = await hasUserSkippedBuggyAppWarning.wait();
+
+        if (response === 'cancelled-operation') {
+          ledgerNavigate.cancelLedgerAction();
+          return;
+        }
+      }
+    }
```

Comment:
How could we pass in this check from the Stacks-specific use of this hook, such that `useLedgerSignTx` knows less about any given chain? Could we pass in as a callback? An optional `afterConnect()` passed in?

Reply:
Done! Used an optional function `passesAdditionalVersionCheck` for that purpose